README.md
# API Problem

[![CI workflow](https://github.com/montealegreluis/api-problem/actions/workflows/ci.yml/badge.svg)](https://github.com/montealegreluis/api-problem/actions/workflows/ci.yml)
[![Release workflow](https://github.com/montealegreluis/api-problem/actions/workflows/release.yml/badge.svg)](https://github.com/montealegreluis/api-problem/actions/workflows/release.yml)
[![semantic-release: conventional-commits](https://img.shields.io/badge/semantic--release-conventionalcommits-e10079?logo=semantic-release)](https://github.com/semantic-release/semantic-release)

API Problem implements [application/problem+json](https://tools.ietf.org/html/rfc7807).
It comes with convenient factories and builders for most common use cases.

## Installation

1. [Authenticating to GitHub Packages](https://github.com/MontealegreLuis/api-problem/blob/main/docs/installation/authentication.md)
2. [Maven](https://github.com/MontealegreLuis/api-problem/blob/main/docs/installation/maven.md)
3. [Gradle](https://github.com/MontealegreLuis/api-problem/blob/main/docs/installation/gradle.md)

## Usage

### Creating problems

There are cases in which an HTTP status code is enough to convey the necessary information. 

```java
ApiProblem.from(Status.NOT_FOUND);
```

Will produce this:

```json
{
 "title": "Not Found",
 "status": 404,
 "type": "https://www.w3.org/Protocols/rfc2616/rfc2616-sec10.html#sec10.4.5"
}
```

As specified by [Predefined Problem Types](https://tools.ietf.org/html/rfc7807#section-4.2):

> The "about:blank" URI, when used as a problem type, indicates that the problem has no additional semantics beyond that of the HTTP status code.

In this case use the following factory.

```java
ApiProblem.witDefaultType(Status.NOT_FOUND);
```

It will produce this:

```json
{
 "title": "Not Found",
 "status": 404
}
```

### Problem Builder

Most of the time you'll need to define problem types that are unique to your application. 
The Problem Builder offers a fluent API to create problem instances without the need to create custom classes:

```java
ApiProblemBuilder.aProblem()
  .withType(URI.create("https://example.org/not-found"))
  .withTitle("Concert not found")
  .withStatus(Status.NOT_FOUND.code())
  .withDetails("Concert with ID 2ca332b9-3f69-4882-abc9-58f534217bdd cannot be found")
  .with("concert", "2ca332b9-3f69-4882-abc9-58f534217bdd")
  .build();
```

Producing the following response.

```json
{
  "type": "https://example.org/not-found",
  "title": "Concert not found",
  "status": 404,
  "details": "Concert with ID 2ca332b9-3f69-4882-abc9-58f534217bdd cannot be found",
  "concert": "2ca332b9-3f69-4882-abc9-58f534217bdd"
}
```

### Custom Problems

The final stage of customization is achieved by sub-classing `ApiProblem`. 

```java
public final class ConcertNotFoundProblem extends ApiProblem {
  private static final URI TYPE = URI.create("https://example.com/not-found");
  private final String concert;

  public ConcertNotFoundProblem(final String concert) {
    super("Concert not found", 404, TYPE, format("Concert with ID %s cannot be found", concert), null, null);
    this.product = product;
  }

  public String getConcert() {
    return concert;
  }
}
```

### Exceptions in Problem responses

It is also possible to include exception information in an API Problem response.

```java
aProblem()
  .from(INTERNAL_SERVER_ERROR)
  .withDetails(exception.getMessage())
  .withException(exception)
  .build());
```

The example above would be represented as JSON as follows.

```json
{
  "type": "https://www.w3.org/Protocols/rfc2616/rfc2616-sec10.html#sec10.5.1",
  "title": "Internal Server Error",
  "status": 500,
  "details": "For input string \"two\"",
  "exception": {
    "message": "For input string \"two\"",
    "class": "java.lang.NumberFormatException",
    "line": 65,
    "file": "NumberFormatException.java",
    "trace": [
      "java.base/java.lang.NumberFormatException.forInputString(NumberFormatException.java:65)",
      "java.base/java.lang.Integer.parseInt(Integer.java:652)",
      "java.base/java.lang.Integer.parseInt(Integer.java:770)",
      "com.montealegreluis.activityfeed.Application.main(Application.java:10)"
    ]
  }
}
```

The response will include previous exceptions if available.

## Contribute

Please refer to [CONTRIBUTING](https://github.com/MontealegreLuis/api-problem/blob/main/CONTRIBUTING.md) for information on how to contribute to API Problem.

## License

Released under the [BSD-3-Clause](https://github.com/MontealegreLuis/api-problem/blob/main/LICENSE).


CHANGELOG.md
# Changelog

All notable changes to API Problem are documented using the [Keep a CHANGELOG](http://keepachangelog.com/) principles.

Please refer to the [releases page](https://github.com/MontealegreLuis/api-problem/releases) for specific details.


src/test/java/com/montealegreluis/apiproblem/StatusTest.java
package com.montealegreluis.apiproblem;

import static org.junit.jupiter.api.Assertions.*;

import java.net.URI;
import org.junit.jupiter.api.Test;

final class StatusTest {
  @Test
  void it_knows_its_reason_phrase_code_and_type() {
    var status = Status.METHOD_NOT_ALLOWED;

    assertEquals(405, status.code());
    assertEquals("Method Not Allowed", status.reason());
    assertEquals(
        URI.create("https://www.w3.org/Protocols/rfc2616/rfc2616-sec10.html#sec10.4.6"),
        status.type());
  }
}


src/test/java/com/montealegreluis/apiproblem/ApiProblemTest.java
package com.montealegreluis.apiproblem;

import static org.junit.jupiter.api.Assertions.*;

import com.montealegreluis.assertions.IllegalArgumentException;
import java.net.URI;
import org.junit.jupiter.api.Test;

final class ApiProblemTest {
  @Test
  void it_can_be_created_from_a_given_status() {
    var problem = ApiProblem.from(Status.NOT_FOUND);

    assertEquals(Status.NOT_FOUND.code(), problem.getStatus());
    assertEquals(Status.NOT_FOUND.type(), problem.getType());
    assertEquals(Status.NOT_FOUND.reason(), problem.getTitle());
  }

  @Test
  void it_can_be_created_with_a_default_type() {
    var problem = ApiProblem.witDefaultType(Status.CONFLICT);

    assertEquals(Status.CONFLICT.code(), problem.getStatus());
    assertEquals(ApiProblem.DEFAULT_TYPE, problem.getType());
    assertEquals(Status.CONFLICT.reason(), problem.getTitle());
  }

  @Test
  void it_prevents_creating_a_problem_without_a_title() {
    assertThrows(
        IllegalArgumentException.class,
        () -> new ApiProblem(null, 404, URI.create("https://example.com"), null, null, null));
  }

  @Test
  void it_prevents_creating_a_problem_without_a_status() {
    assertThrows(
        IllegalArgumentException.class,
        () ->
            new ApiProblem("Not Found", null, URI.create("https://example.com"), null, null, null));
  }

  @Test
  void it_prevents_creating_a_problem_without_a_type() {
    assertThrows(
        IllegalArgumentException.class,
        () -> new ApiProblem("Not Found", 404, null, null, null, null));
  }

  @Test
  void it_can_be_compared_to_another_problem() {
    var problem1 = ApiProblem.from(Status.FORBIDDEN);
    var problem2 = ApiProblem.witDefaultType(Status.FORBIDDEN);
    var problem3 =
        ApiProblemBuilder.aProblem()
            .withType(URI.create("https://example.org/not-found"))
            .withTitle("Concert not found")
            .withStatus(Status.NOT_FOUND.code())
            .withDetail("Concert with ID 2ca332b9-3f69-4882-abc9-58f534217bdd cannot be found")
            .with("concert", "2ca332b9-3f69-4882-abc9-58f534217bdd")
            .build();

    assertEquals(problem1, problem1);
    assertNotEquals(problem1, problem2);
    assertNotEquals(problem1, null);
    assertNotEquals(problem2, problem3);
  }
}


src/test/java/com/montealegreluis/apiproblem/ApiProblemBuilderTest.java
package com.montealegreluis.apiproblem;

import static com.montealegreluis.apiproblem.ApiProblemBuilder.aProblem;
import static com.montealegreluis.apiproblem.Status.*;
import static org.junit.jupiter.api.Assertions.*;
import static org.junit.jupiter.api.Assertions.assertEquals;

import com.montealegreluis.assertions.IllegalArgumentException;
import java.net.URI;
import java.util.stream.Stream;
import org.junit.jupiter.api.DisplayName;
import org.junit.jupiter.api.Test;
import org.junit.jupiter.params.ParameterizedTest;
import org.junit.jupiter.params.provider.Arguments;
import org.junit.jupiter.params.provider.MethodSource;

final class ApiProblemBuilderTest {
  @Test
  void it_prevents_creating_a_problem_without_a_title() {
    assertThrows(IllegalArgumentException.class, () -> aProblem().build());
  }

  @Test
  void it_prevents_creating_a_problem_without_a_status() {
    assertThrows(IllegalArgumentException.class, () -> aProblem().withTitle("Not Found").build());
  }

  @Test
  void it_prevents_creating_a_problem_without_a_type() {
    assertThrows(
        IllegalArgumentException.class,
        () -> aProblem().withTitle("Not Found").withStatus(404).build());
  }

  @Test
  void it_creates_a_problem_with_minimum_information() {
    var problem = ApiProblem.from(INTERNAL_SERVER_ERROR);

    assertEquals(
        problem,
        aProblem()
            .withTitle(INTERNAL_SERVER_ERROR.reason())
            .withStatus(INTERNAL_SERVER_ERROR.code())
            .withType(INTERNAL_SERVER_ERROR.type())
            .build());
  }

  @Test
  void it_creates_a_problem_with_details() {
    var details = "Cannot find item, 123 is not a valid identifier";

    var problem =
        aProblem()
            .withTitle(UNPROCESSABLE_ENTITY.reason())
            .withStatus(UNPROCESSABLE_ENTITY.code())
            .withType(UNPROCESSABLE_ENTITY.type())
            .withDetail(details)
            .build();

    assertEquals(UNPROCESSABLE_ENTITY.reason(), problem.getTitle());
    assertEquals(UNPROCESSABLE_ENTITY.code(), problem.getStatus());
    assertEquals(UNPROCESSABLE_ENTITY.type(), problem.getType());
    assertEquals(details, problem.getDetail());
  }

  @Test
  void it_prevents_creating_a_problem_with_blank_or_null_details() {
    assertThrows(
        IllegalArgumentException.class,
        () ->
            aProblem()
                .withTitle(UNPROCESSABLE_ENTITY.reason())
                .withStatus(UNPROCESSABLE_ENTITY.code())
                .withType(UNPROCESSABLE_ENTITY.type())
                .withDetail(" ")
                .build());

    assertThrows(
        IllegalArgumentException.class,
        () ->
            aProblem()
                .withTitle(UNPROCESSABLE_ENTITY.reason())
                .withStatus(UNPROCESSABLE_ENTITY.code())
                .withType(UNPROCESSABLE_ENTITY.type())
                .withDetail(null)
                .build());
  }

  @Test
  void it_creates_a_problem_with_instance() {
    var instance = URI.create("https://example.com/problem");

    var problem =
        aProblem()
            .withTitle(BAD_REQUEST.reason())
            .withStatus(BAD_REQUEST.code())
            .withType(BAD_REQUEST.type())
            .withInstance(instance)
            .build();

    assertEquals(BAD_REQUEST.reason(), problem.getTitle());
    assertEquals(BAD_REQUEST.code(), problem.getStatus());
    assertEquals(BAD_REQUEST.type(), problem.getType());
    assertEquals(instance, problem.getInstance());
  }

  @Test
  void it_prevents_creating_a_problem_with_a_null_instance() {
    assertThrows(
        IllegalArgumentException.class,
        () ->
            aProblem()
                .withTitle(UNPROCESSABLE_ENTITY.reason())
                .withStatus(UNPROCESSABLE_ENTITY.code())
                .withType(UNPROCESSABLE_ENTITY.type())
                .withInstance(null)
                .build());
  }

  @Test
  void it_creates_a_problem_with_additional_properties() {
    var username = "jane.doe";

    var problem =
        aProblem()
            .withTitle(PRECONDITION_FAILED.reason())
            .withStatus(PRECONDITION_FAILED.code())
            .withType(PRECONDITION_FAILED.type())
            .with("username", username)
            .build();

    assertEquals(PRECONDITION_FAILED.reason(), problem.getTitle());
    assertEquals(PRECONDITION_FAILED.code(), problem.getStatus());
    assertEquals(PRECONDITION_FAILED.type(), problem.getType());
    assertEquals(1, problem.getAdditionalProperties().size());
    assertTrue(problem.getAdditionalProperties().containsKey("username"));
    assertEquals(username, problem.getAdditionalProperties().get("username"));
  }

  @DisplayName("it prevents adding properties_to_a_problem_using_reserved_properties")
  @ParameterizedTest(name = "property={0}, value={1}")
  @MethodSource("propertiesProvider")
  void it_prevents_adding_properties_to_a_problem_using_reserved_properties(
      String propertyName, Object propertyValue) {
    assertThrows(
        IllegalArgumentException.class,
        () ->
            aProblem()
                .withTitle(UNPROCESSABLE_ENTITY.reason())
                .withStatus(UNPROCESSABLE_ENTITY.code())
                .withType(UNPROCESSABLE_ENTITY.type())
                .with(propertyName, propertyValue)
                .build());
  }

  private static Stream<Arguments> propertiesProvider() {
    return Stream.of(
        Arguments.of("type", URI.create("https://example.com/problem")),
        Arguments.of("title", "Unprocessable Entity"),
        Arguments.of("status", 422),
        Arguments.of("detail", "Input provided is invalid"),
        Arguments.of("instance", URI.create("https://example.com/instance")));
  }

  @Test
  void it_creates_a_problem_with_an_exception() {
    var exception = new RuntimeException("Something went wrong");

    var problem =
        aProblem()
            .withTitle(BAD_REQUEST.reason())
            .withStatus(BAD_REQUEST.code())
            .withType(BAD_REQUEST.type())
            .withException(exception)
            .build();

    assertEquals(BAD_REQUEST.reason(), problem.getTitle());
    assertEquals(BAD_REQUEST.code(), problem.getStatus());
    assertEquals(BAD_REQUEST.type(), problem.getType());
    assertEquals(1, problem.getAdditionalProperties().size());
    assertTrue(problem.getAdditionalProperties().containsKey("exception"));
  }

  @Test
  void it_prevents_creating_a_problem_with_a_null_exception() {
    assertThrows(
        IllegalArgumentException.class,
        () ->
            aProblem()
                .withTitle(BAD_REQUEST.reason())
                .withStatus(BAD_REQUEST.code())
                .withType(BAD_REQUEST.type())
                .withException(null)
                .build());
  }

  @Test
  void it_creates_a_problem_from_a_given_status() {
    var problem = aProblem().from(FORBIDDEN).build();

    assertEquals(FORBIDDEN.reason(), problem.getTitle());
    assertEquals(FORBIDDEN.code(), problem.getStatus());
    assertEquals(FORBIDDEN.type(), problem.getType());
  }

  @Test
  void it_creates_a_problem_from_a_given_status_with_default_type() {
    var problem = aProblem().witDefaultType(UNAUTHORIZED).build();

    assertEquals(UNAUTHORIZED.reason(), problem.getTitle());
    assertEquals(UNAUTHORIZED.code(), problem.getStatus());
    assertEquals(URI.create("about:blank"), problem.getType());
  }
}


src/main/java/com/montealegreluis/apiproblem/Problem.java
package com.montealegreluis.apiproblem;

import java.net.URI;
import java.util.Map;

public interface Problem {
  URI DEFAULT_TYPE = URI.create("about:blank");

  /**
   * An absolute URI that identifies the problem type.
   *
   * <p>When this member is not present, its value is assumed to be "about:blank".
   */
  URI getType();

  /**
   * A short, human-readable summary of the problem type.
   *
   * <p>It SHOULD NOT change from occurrence to occurrence of the problem.
   */
  String getTitle();

  /** The HTTP status code generated by the origin server for this occurrence of the problem. */
  Integer getStatus();

  /** A human-readable explanation specific to this occurrence of the problem. */
  String getDetail();

  /** An absolute URI that identifies the specific occurrence of the problem. */
  URI getInstance();

  /**
   * Optional additional attributes of the problem.
   *
   * <p>Implementations can choose to ignore this in favor of concrete typed fields.
   */
  Map<String, Object> getAdditionalProperties();
}


src/main/java/com/montealegreluis/apiproblem/ApiProblem.java
package com.montealegreluis.apiproblem;

import com.montealegreluis.assertions.Assert;
import java.net.URI;
import java.util.Collections;
import java.util.LinkedHashMap;
import java.util.Map;
import java.util.Optional;
import lombok.EqualsAndHashCode;

@EqualsAndHashCode
public class ApiProblem implements Problem {
  private final String title;
  private final Integer status;
  private final URI type;
  private final String detail;
  private final URI instance;
  private final Map<String, Object> additionalProperties;

  public static ApiProblem from(Status status) {
    return new ApiProblem(status.reason(), status.code(), status.type(), null, null, null);
  }

  public static ApiProblem witDefaultType(Status status) {
    return new ApiProblem(status.reason(), status.code(), DEFAULT_TYPE, null, null, null);
  }

  public ApiProblem(
      String title,
      Integer status,
      URI type,
      String detail,
      URI instance,
      Map<String, Object> additionalProperties) {
    Assert.notBlank(title, "Title cannot be blank or null");
    this.title = title;
    Assert.notNull(status, "Status cannot be null");
    this.status = status;
    Assert.notNull(type, "Type cannot be null");
    this.type = type;
    this.detail = detail;
    this.instance = instance;
    this.additionalProperties =
        Optional.ofNullable(additionalProperties).orElseGet(LinkedHashMap::new);
  }

  @Override
  public URI getType() {
    return type;
  }

  @Override
  public String getTitle() {
    return title;
  }

  @Override
  public Integer getStatus() {
    return status;
  }

  @Override
  public String getDetail() {
    return detail;
  }

  @Override
  public URI getInstance() {
    return instance;
  }

  @Override
  public Map<String, Object> getAdditionalProperties() {
    return Collections.unmodifiableMap(additionalProperties);
  }
}


src/main/java/com/montealegreluis/apiproblem/ApiProblemBuilder.java
package com.montealegreluis.apiproblem;

import static com.montealegreluis.activityfeed.ExceptionContextFactory.contextFrom;

import com.montealegreluis.assertions.Assert;
import java.net.URI;
import java.util.LinkedHashMap;
import java.util.List;
import java.util.Map;
import lombok.AccessLevel;
import lombok.NoArgsConstructor;

@NoArgsConstructor(access = AccessLevel.PROTECTED)
public class ApiProblemBuilder {
  protected static final List<String> RESERVED_PROPERTIES =
      List.of("type", "title", "status", "detail", "instance");
  protected String title;
  protected Integer status;
  protected URI type;
  protected String detail;
  protected URI instance;
  protected final Map<String, Object> additionalProperties = new LinkedHashMap<>();

  public static ApiProblemBuilder aProblem() {
    return new ApiProblemBuilder();
  }

  public ApiProblemBuilder from(Status status) {
    title = status.reason();
    this.status = status.code();
    type = status.type();
    return this;
  }

  public ApiProblemBuilder witDefaultType(Status status) {
    title = status.reason();
    this.status = status.code();
    type = Problem.DEFAULT_TYPE;
    return this;
  }

  public ApiProblemBuilder withTitle(String title) {
    this.title = title;
    return this;
  }

  public ApiProblemBuilder withStatus(Integer status) {
    this.status = status;
    return this;
  }

  public ApiProblemBuilder withType(URI type) {
    this.type = type;
    return this;
  }

  public ApiProblemBuilder withDetail(String detail) {
    Assert.notBlank(detail, "Detail cannot be blank or null");
    this.detail = detail;
    return this;
  }

  public ApiProblemBuilder withInstance(URI instance) {
    Assert.notNull(instance, "Instance cannot be null");
    this.instance = instance;
    return this;
  }

  public ApiProblemBuilder withException(Throwable exception) {
    Assert.notNull(exception, "Exception cannot be null");
    additionalProperties.put("exception", contextFrom(exception));
    return this;
  }

  public ApiProblemBuilder with(String additionalProperty, Object value) {
    Assert.notIn(additionalProperty, RESERVED_PROPERTIES, "%s is a reserved property");
    additionalProperties.put(additionalProperty, value);
    return this;
  }

  public ApiProblem build() {
    return new ApiProblem(title, status, type, detail, instance, additionalProperties);
  }
}


src/main/java/com/montealegreluis/apiproblem/Status.java
package com.montealegreluis.apiproblem;

import java.net.URI;

public enum Status {
  CONTINUE(100, "Continue", URI.create("https://tools.ietf.org/html/rfc7231#section-6.2.1")),
  SWITCHING_PROTOCOLS(
      101, "Switching Protocols", URI.create("https://tools.ietf.org/html/rfc7231#section-6.2.2")),
  PROCESSING(102, "Processing", URI.create("https://tools.ietf.org/html/rfc2518#section-10.1")),
  CHECKPOINT(
      103,
      "Checkpoint",
      URI.create("https://code.google.com/p/gears/wiki/ResumableHttpRequestsProposal")),
  OK(200, "OK", URI.create("https://www.w3.org/Protocols/rfc2616/rfc2616-sec10.html#sec10.2.1")),
  CREATED(
      201,
      "Created",
      URI.create("https://www.w3.org/Protocols/rfc2616/rfc2616-sec10.html#sec10.2.2")),
  ACCEPTED(
      202,
      "Accepted",
      URI.create("http://www.w3.org/Protocols/rfc2616/rfc2616-sec10.html#sec10.2.3")),
  NON_AUTHORITATIVE_INFORMATION(
      203,
      "Non-Authoritative Information",
      URI.create("https://tools.ietf.org/html/rfc7231#section-6.3.4")),
  NO_CONTENT(
      204,
      "No Content",
      URI.create("https://www.w3.org/Protocols/rfc2616/rfc2616-sec10.html#sec10.2.5")),
  RESET_CONTENT(
      205,
      "Reset Content",
      URI.create("https://www.w3.org/Protocols/rfc2616/rfc2616-sec10.html#sec10.2.6")),
  PARTIAL_CONTENT(
      206,
      "Partial Content",
      URI.create("https://www.w3.org/Protocols/rfc2616/rfc2616-sec10.html#sec10.2.7")),
  MULTI_STATUS(207, "Multi-Status", URI.create("https://tools.ietf.org/html/rfc4918#section-13")),
  ALREADY_REPORTED(
      208, "Already Reported", URI.create("https://tools.ietf.org/html/rfc5842#section-7.1")),
  IM_USED(226, "IM Used", URI.create("https://tools.ietf.org/html/rfc3229#section-10.4.1")),
  MULTIPLE_CHOICES(
      300, "Multiple Choices", URI.create("https://tools.ietf.org/html/rfc7231#section-6.4.1")),
  MOVED_PERMANENTLY(
      301,
      "Moved Permanently",
      URI.create("https://www.w3.org/Protocols/rfc2616/rfc2616-sec10.html#sec10.3.2")),
  FOUND(
      302,
      "Found",
      URI.create("https://www.w3.org/Protocols/rfc2616/rfc2616-sec10.html#sec10.3.3")),
  SEE_OTHER(
      303,
      "See Other",
      URI.create("https://www.w3.org/Protocols/rfc2616/rfc2616-sec10.html#sec10.3.4")),
  NOT_MODIFIED(
      304,
      "Not Modified",
      URI.create("https://www.w3.org/Protocols/rfc2616/rfc2616-sec10.html#sec10.3.5")),
  USE_PROXY(
      305,
      "Use Proxy",
      URI.create("https://www.w3.org/Protocols/rfc2616/rfc2616-sec10.html#sec10.3.6")),
  TEMPORARY_REDIRECT(
      307,
      "Temporary Redirect",
      URI.create("https://www.w3.org/Protocols/rfc2616/rfc2616-sec10.html#sec10.3.8")),
  PERMANENT_REDIRECT(308, "Permanent Redirect", URI.create("https://tools.ietf.org/html/rfc7238")),
  BAD_REQUEST(
      400,
      "Bad Request",
      URI.create("https://www.w3.org/Protocols/rfc2616/rfc2616-sec10.html#sec10.4.1")),
  UNAUTHORIZED(
      401,
      "Unauthorized",
      URI.create("https://www.w3.org/Protocols/rfc2616/rfc2616-sec10.html#sec10.4.2")),
  PAYMENT_REQUIRED(
      402,
      "Payment Required",
      URI.create("https://www.w3.org/Protocols/rfc2616/rfc2616-sec10.html#sec10.4.3")),
  FORBIDDEN(
      403,
      "Forbidden",
      URI.create("https://www.w3.org/Protocols/rfc2616/rfc2616-sec10.html#sec10.4.4")),
  NOT_FOUND(
      404,
      "Not Found",
      URI.create("https://www.w3.org/Protocols/rfc2616/rfc2616-sec10.html#sec10.4.5")),
  METHOD_NOT_ALLOWED(
      405,
      "Method Not Allowed",
      URI.create("https://www.w3.org/Protocols/rfc2616/rfc2616-sec10.html#sec10.4.6")),
  NOT_ACCEPTABLE(
      406,
      "Not Acceptable",
      URI.create("https://www.w3.org/Protocols/rfc2616/rfc2616-sec10.html#sec10.4.7")),
  PROXY_AUTHENTICATION_REQUIRED(
      407,
      "Proxy Authentication Required",
      URI.create("https://www.w3.org/Protocols/rfc2616/rfc2616-sec10.html#sec10.4.8")),
  REQUEST_TIMEOUT(
      408,
      "Request Timeout",
      URI.create("https://www.w3.org/Protocols/rfc2616/rfc2616-sec10.html#sec10.4.9")),
  CONFLICT(
      409,
      "Conflict",
      URI.create("https://www.w3.org/Protocols/rfc2616/rfc2616-sec10.html#sec10.4.10")),
  GONE(
      410,
      "Gone",
      URI.create("https://www.w3.org/Protocols/rfc2616/rfc2616-sec10.html#sec10.4.11")),
  LENGTH_REQUIRED(
      411,
      "Length Required",
      URI.create("https://www.w3.org/Protocols/rfc2616/rfc2616-sec10.html#sec10.4.12")),
  PRECONDITION_FAILED(
      412,
      "Precondition Failed",
      URI.create("https://www.w3.org/Protocols/rfc2616/rfc2616-sec10.html#sec10.4.13")),
  REQUEST_ENTITY_TOO_LARGE(
      413,
      "Request Entity Too Large",
      URI.create("https://www.w3.org/Protocols/rfc2616/rfc2616-sec10.html#sec10.4.14")),
  REQUEST_URI_TOO_LONG(
      414,
      "Request-URI Too Long",
      URI.create("https://www.w3.org/Protocols/rfc2616/rfc2616-sec10.html#sec10.4.15")),
  UNSUPPORTED_MEDIA_TYPE(
      415,
      "Unsupported Media Type",
      URI.create("https://www.w3.org/Protocols/rfc2616/rfc2616-sec10.html#sec10.4.16")),
  REQUESTED_RANGE_NOT_SATISFIABLE(
      416,
      "Requested Range Not Satisfiable",
      URI.create("https://www.w3.org/Protocols/rfc2616/rfc2616-sec10.html#sec10.4.17")),
  EXPECTATION_FAILED(
      417,
      "Expectation Failed",
      URI.create("https://www.w3.org/Protocols/rfc2616/rfc2616-sec10.html#sec10.4.18")),
  I_AM_A_TEAPOT(
      418, "I'm a teapot", URI.create("https://tools.ietf.org/html/rfc2324#section-2.3.2")),
  UNPROCESSABLE_ENTITY(
      422, "Unprocessable Entity", URI.create("https://tools.ietf.org/html/rfc4918#section-11.2")),
  LOCKED(423, "Locked", URI.create("https://tools.ietf.org/html/rfc4918#section-11.3")),
  FAILED_DEPENDENCY(
      424, "Failed Dependency", URI.create("https://tools.ietf.org/html/rfc4918#section-11.4")),
  UPGRADE_REQUIRED(
      426, "Upgrade Required", URI.create("https://tools.ietf.org/html/rfc2817#section-6")),
  PRECONDITION_REQUIRED(
      428, "Precondition Required", URI.create("https://tools.ietf.org/html/rfc6585#section-3")),
  TOO_MANY_REQUESTS(
      429, "Too Many Requests", URI.create("https://tools.ietf.org/html/rfc6585#section-4")),
  REQUEST_HEADER_FIELDS_TOO_LARGE(
      431,
      "Request Header Fields Too Large",
      URI.create("https://tools.ietf.org/html/rfc6585#section-5")),
  UNAVAILABLE_FOR_LEGAL_REASONS(
      451,
      "Unavailable For Legal Reasons",
      URI.create("https://tools.ietf.org/html/rfc7725#section-3")),
  INTERNAL_SERVER_ERROR(
      500,
      "Internal Server Error",
      URI.create("https://www.w3.org/Protocols/rfc2616/rfc2616-sec10.html#sec10.5.1")),
  NOT_IMPLEMENTED(
      501,
      "Not Implemented",
      URI.create("https://www.w3.org/Protocols/rfc2616/rfc2616-sec10.html#sec10.5.2")),
  BAD_GATEWAY(
      502,
      "Bad Gateway",
      URI.create("https://www.w3.org/Protocols/rfc2616/rfc2616-sec10.html#sec10.5.3")),
  SERVICE_UNAVAILABLE(
      503,
      "Service Unavailable",
      URI.create("https://www.w3.org/Protocols/rfc2616/rfc2616-sec10.html#sec10.5.4")),
  GATEWAY_TIMEOUT(
      504,
      "Gateway Timeout",
      URI.create("https://www.w3.org/Protocols/rfc2616/rfc2616-sec10.html#sec10.5.5")),
  HTTP_VERSION_NOT_SUPPORTED(
      505,
      "HTTP Version Not Supported",
      URI.create("https://www.w3.org/Protocols/rfc2616/rfc2616-sec10.html#sec10.5.6")),
  VARIANT_ALSO_NEGOTIATES(
      506,
      "Variant Also Negotiates",
      URI.create("https://tools.ietf.org/html/rfc2295#section-8.1")),
  INSUFFICIENT_STORAGE(
      507, "Insufficient Storage", URI.create("https://tools.ietf.org/html/rfc4918#section-11.5")),
  LOOP_DETECTED(
      508, "Loop Detected", URI.create("https://tools.ietf.org/html/rfc5842#section-7.2")),
  BANDWIDTH_LIMIT_EXCEEDED(509, "Bandwidth Limit Exceeded", URI.create("about:blank")),
  NOT_EXTENDED(510, "Not Extended", URI.create("https://tools.ietf.org/html/rfc2774#section-7")),
  NETWORK_AUTHENTICATION_REQUIRED(
      511,
      "Network Authentication Required",
      URI.create("https://tools.ietf.org/html/rfc6585#section-6"));

  private final int code;
  private final String reason;
  private final URI type;

  Status(int code, String reason, URI type) {
    this.code = code;
    this.reason = reason;
    this.type = type;
  }

  public int code() {
    return code;
  }

  public String reason() {
    return reason;
  }

  public URI type() {
    return type;
  }
}
